packages/ai-constructs/src/conversation/runtime/bedrock_converse_adapter.ts

```diff
@@ -3,43 +3,156 @@ import {
   ContentBlock,
   ConverseCommand,
   ConverseCommandInput,
+  ConverseCommandOutput,
   Message,
+  Tool,
+  ToolConfiguration,
 } from '@aws-sdk/client-bedrock-runtime';
-import { ConversationTurnEvent } from './types.js';
+import { ConversationTurnEvent, ExecutableTool } from './types.js';
+import { ConversationTurnEventToolsProvider } from './event-tools-provider';
 
 /**
  * This class is responsible for interacting with Bedrock Converse API
  * in order to produce final response that can be sent back to caller.
  */
 export class BedrockConverseAdapter {
+  private readonly allTools: Array<ExecutableTool>;
+  private readonly toolByName: Map<string, ExecutableTool> = new Map();
+
   /**
    * Creates Bedrock Converse Adapter.
    */
   constructor(
     private readonly event: ConversationTurnEvent,
+    additionalTools: Array<ExecutableTool>,
     private readonly bedrockClient: BedrockRuntimeClient = new BedrockRuntimeClient(
       { region: event.modelConfiguration.region }
-    )
-  ) {}
+    ),
+    eventToolsProvider = new ConversationTurnEventToolsProvider(event)
+  ) {
+    this.allTools = [...eventToolsProvider.getEventTools(), ...additionalTools];
+    this.allTools.forEach((t) => {
+      if (this.toolByName.has(t.name)) {
+        throw new Error(
+          `Tools must have unique names. Duplicate tools with '${t.name}' name detected.`
+        );
```

Comment:
Nit: I think it would be a better experience to collect all duplicate names and then throw

Reply:
Done.